### PR TITLE
feat(notifications): Hivemind has saved you ~Nk tokens recap

### PR DESCRIPTION
**Pull request by kaghni**
## Summary

Per-session savings notification at session start using the published LoCoMo benchmark multiplier:

```
🐝 Hivemind has saved you ~12.9k tokens
   14 sessions · 63 memory searches
```

## Formula (plain math, X − Y = Z)

```
Y = memorySearchBytes / 4
    bytes returned from Bash tool calls grep'ing ~/.deeplake/memory/
    (memory summaries + sessions raw — both live under that root),
    summed across all records in ~/.deeplake/usage-stats.jsonl,
    divided by 4 to approximate tokens (BPE rule-of-thumb)

X = 1.7 · Y
    counterfactual: tokens to do the same work without hivemind,
    per the published LoCoMo benchmark at deeplake.ai/hivemind —
    1,008 vs 1,700 tokens / Q with Claude Haiku via `claude -p`,
    hybrid lexical + semantic retrieval

Z = X − Y = 0.7 · Y
    tokens saved (the headline number)
```

No LLM calls, no SQL, no new backend endpoints. Pure local file IO + arithmetic.

## What's new

| File | Purpose |
|---|---|
| `src/notifications/transcript-parser.ts` | Walk session JSONL at SessionEnd, extract `memorySearchBytes` (bytes Claude received from memory-grep tool calls — matched via `tool_use_id` to `tool_result` content) and `memorySearchCount`. |
| `src/notifications/usage-tracker.ts` | Append/read `~/.deeplake/usage-stats.jsonl`. Minimal record shape `{endedAt, sessionId, memorySearchBytes, memorySearchCount}`. Backward-compat reader for legacy records. |
| `src/notifications/sources/local-usage.ts` | At SessionStart, sum across all records, apply `Z = 0.7 · Y / 4`, render. dedupKey = `{session: sessionId}` so each new session refires with updated numbers. |

## What's modified

- `src/notifications/index.ts` — thread sessionId from hook stdin, wire `fetchLocalUsageNotifications` into the source list, filter through `tryClaim` to prevent duplicate emission from the dual SessionStart hook registration (settings.json + marketplace `hooks.json` both fire in parallel).
- `src/notifications/state.ts` — `tryClaim()` uses `openSync(path, "wx")` (O_CREAT|O_EXCL semantics) on per-(id, dedupKey-hash) claim files. First process wins, second gets EEXIST and skips. Fail-open on non-EEXIST FS errors.
- `src/hooks/session-end.ts` — parse `transcript_path` from stdin, call `recordSessionUsage()` before the wiki-worker lock check.
- `src/hooks/session-notifications.ts` — extract `session_id` from stdin (previously discarded).
- `src/cli/install-claude.ts` — `syncHivemindHooksToSettings()`: rewrites hivemind-owned hook entries in `~/.claude/settings.json` from the marketplace's canonical `hooks.json` on every install/update. Fixes a real install bug — users whose first install predated `session-notifications.js` weren't seeing it fire after `hivemind update` because settings.json was never re-synced.

## What's deferred

- **Channel 3 (skill auto-pull bytes in Y)** — filed as #127. The current Y is a conservative lower bound; adding skill bytes will jump the displayed number.
- **Channel 4 (SessionStart instructions block)** — explicitly excluded; it's hivemind explaining itself, not user past content.
- **$-figure rendering** — model pricing varies; skip for v1.

## Test plan

- [x] Unit tests: 2263 / 2263 passing
- [x] Per-file coverage thresholds met
  - `transcript-parser.ts`: 90/75/90/90
  - `usage-tracker.ts`: 90/75/90/90
  - `sources/local-usage.ts`: 90/80/90/90
  - `state.ts`: 90/75/90/90 (branches bumped from 80 for fail-open paths)
- [x] Live E2E on test machine:
  - Single hook fires renders `🐝 Hivemind has saved you ~12.9k tokens / 14 sessions · 63 memory searches`
  - Concurrent race (2 parallel processes, same session_id): exactly ONE emission. The other process's stdout is empty (claim file held).
  - New `session_id`: refires with cumulative numbers.
  - Re-invoke same session: silent (state dedup).
- [x] Backward-compat verified: legacy records (predating `memorySearchCount`) are now counted in "N sessions" instead of silently dropped.

## Verification math (live numbers from test machine)

```
total memorySearchBytes across 14 records = 73,474
Y = 73,474 / 4 = 18,368 tokens delivered
Z = 0.7 × 18,368 = 12,858 tokens saved → rounds to ~12.9k ✓
memorySearchCount total = 63 → "63 memory searches" ✓
records = 14 → "14 sessions" ✓
```

## Citation

LoCoMo benchmark at deeplake.ai/hivemind: 100 QA, Claude Haiku via `claude -p`, hybrid lexical + semantic retrieval, **1,008 vs 1,700 tokens / Q** → ratio 1.687 → rounded to 1.7×. Also: 25% cheaper ($6.65 vs $8.94 / 100 QA), 31% fewer turns (6.2 vs 8.9).

## Summary by CodeRabbit

* **New Features**
  * Per-session memory usage tracking records and automatically reports lookup activity metrics for sessions
  * Token savings recap notifications display efficiency improvements achieved through hivemind memory integration
  * Improved plugin installation with enhanced configuration management and hook synchronization capabilities
  * Streamlined and reliable installation workflows across multiple plugins and system integrations

[Review Change Stack](https://app.coderabbit.ai/change-stack/activeloopai/hivemind/pull/128)

### CODE REVIEW

File: bundle/cli.js [hunk 1]
```diff
@@ -178,6 +181,70 @@ function pluginAlreadyInstalled() {
...
+}
+function settingsJsonPath() {
+  return join2(homedir2(), ".claude", "settings.json");
+}
+function resolveCommand(command, pluginRoot) {
+  return command.replace(/\$\{CLAUDE_PLUGIN_ROOT\}/g, pluginRoot);
+}
+function isHivemindMatcher(matcher) {
+  return matcher.hooks?.some((h) => typeof h.command === "string" && h.command.includes("plugins/hivemind/bundle/")) ?? false;
+}
```
[the remark this answers is not in the document]
Comment (kaghni) [reply bundle/cli.js:203]: Fixed in 0eecd7f. `isHivemindMatcher()` now normalizes backslashes to forward slashes before the substring check, so a Windows-style `plugins\hivemind\bundle\...` entry is correctly recognized as hivemind-owned and gets replaced rather than preserved alongside the canonical one. Regression test added in `cli-install-claude.test.ts` — seeds settings.json with a Windows-path entry and asserts exactly ONE hook entry survives the sync (not two). Thanks for the catch — would've duplicated hooks for every Windows user.

File: src/hooks/session-notifications.ts [hunk 2]
```diff
@@ -33,12 +33,15 @@ async function main(): Promise<void> {
...
 
-  // Drain stdin so Claude Code's writer doesn't EPIPE; we don't currently
-  // use the input, but future rules may key dedup on session_id.
-  await readStdin<SessionStartInput>().catch(() => ({}));
+  // Drain stdin so Claude Code's writer doesn't EPIPE. We also extract
+  // session_id so the local-usage source can scope its dedupKey to the
+  // current session — two parallel hook fires for the same session share
+  // the same id and dedupe to one emission via the atomic claim file.
+  const input = await readStdin<SessionStartInput>().catch(() => ({} as SessionStartInput));
+  const sessionId = typeof input?.session_id === "string" ? input.session_id : undefined;
```
[the remark this answers is not in the document]
Comment (kaghni) [reply src/hooks/session-notifications.ts]: Fixed in 0eecd7f. session_id now goes through `trim()` + non-empty check before being threaded to drainSessionStart; an empty or whitespace-only value routes to `undefined`, which `fetchLocalUsageNotifications` already handles correctly by returning []. 3 regression tests in `session-notifications-hook.test.ts` cover: empty string → undefined, whitespace-only → undefined, real UUID passes through verbatim. Edge-case in practice (Claude Code always sends a UUID), but defensive and cheap.

**Comment by efenocchi** [comment]
@coderabbitai
Hi

File: src/cli/install-claude.ts [hunk 3]
```diff
@@ -72,6 +75,143 @@ function pluginAlreadyInstalled(): boolean {
...
+  // install would carry `plugins\hivemind\bundle\...` which the POSIX-style
+  // fragment misses. Without this normalization a Windows user's existing
+  // hivemind entries would NOT match, leaving them in `preserved` while
+  // the canonical entries get appended too — duplicating hook registration
+  // on every install/update. (Caught by CodeRabbit, PR #128.)
+  return matcher.hooks?.some(h => {
+    if (typeof h.command !== "string") return false;
+    const normalized = h.command.replace(/\\/g, "/");
+    return normalized.includes("plugins/hivemind/bundle/");
+  }) ?? false;
```
Comment (efenocchi) [inline src/cli/install-claude.ts:153]: @coderabbit what's that?

File: src/cli/install-claude.ts [hunk 4]
```diff
@@ -113,6 +253,20 @@ export function installClaude(): void {
 
   // enable is idempotent in claude CLI — safe to run unconditionally
   runClaude(["plugin", "enable", PLUGIN_KEY]);
+
+  // Sync the plugin's canonical hooks.json into the user's settings.json.
+  // `claude plugin install` writes settings.json the first time, but
+  // `claude plugin update` does NOT re-sync hooks on version bumps, so
+  // users who installed before a hook was added end up never firing it.
+  // This makes the install/update path idempotent w.r.t. hook registration.
+  try {
```
Comment (efenocchi) [inline src/cli/install-claude.ts:262]: @coderabbit explain this

File: src/notifications/sources/local-usage.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+/** 1234 → "1.2k", 12345 → "12.3k", 1234567 → "1.2M". Caller prepends `~`. */
+export function formatTokens(n: number): string {
+  if (!Number.isFinite(n) || n <= 0) return "0";
+  if (n < 1000) return `${Math.round(n)}`;
+  if (n < 100000) return `${(n / 1000).toFixed(1)}k`;
+  if (n < 1000000) return `${Math.round(n / 1000)}k`;
+  return `${(n / 1000000).toFixed(1)}M`;
+}
+
```
Comment (efenocchi) [inline src/notifications/sources/local-usage.ts:55]: @coderabbit why do we have this logic?

File: src/notifications/sources/local-usage.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  if (records.length === 0) {
+    log("no usage records yet — skipping recap");
+    return [];
+  }
+
+  const memorySearchBytes = sumMetric(records, "memorySearchBytes");
+  if (memorySearchBytes <= 0) {
+    log("memorySearchBytes total is 0 — skipping recap");
+    return [];
+  }
```
Comment (efenocchi) [inline src/notifications/sources/local-usage.ts:85]: @coderabbit why do we check bytes instead of token?

File: src/notifications/usage-tracker.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+      // Require the entry to END with `--<userName>` AND have something
+      // before the `--` (not a bare `--kamo`). Stricter than .endsWith()
+      // because a hypothetical user named `--kamo` would otherwise match
+      // an entry literally `--kamo`.
+      const idx = name.lastIndexOf(suffix);
+      if (idx > 0 && idx + suffix.length === name.length) count += 1;
+    }
+    return count;
+  } catch (e: any) {
+    log(`countUserGeneratedSkills readdir failed: ${e?.message ?? String(e)}`);
```
Comment (efenocchi) [inline src/notifications/usage-tracker.ts:166]: @coderabbit in this way if you generate 100 skills and you remove 100skills you will have the same message betweend 2 sessions. should we keep track of merge, deletion, etc?

File: src/notifications/transcript-parser.ts [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+      }
+    }
+    return n;
+  }
+  try {
+    return Buffer.byteLength(JSON.stringify(content ?? ""), "utf-8");
+  } catch {
+    return 0;
+  }
+}
```
Comment (efenocchi) [inline src/notifications/transcript-parser.ts:191]: @coderabbit for what do we use the bytes?

Comment (efenocchi) [reply src/notifications/transcript-parser.ts:191]: @coderabbit why bytes->tokens? can't we keep the tokens? character -> tokens?
is byte -> token better?

**Review by efenocchi — APPROVED** [review]
